### PR TITLE
HG-521 Integrate performance monitoring into Ember & persist with Weppy

### PR DESCRIPTION
**Pull request by kenkouot**

### CODE REVIEW

File: front/scripts/mercury/modules/Trackers/Perf.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+				sample: Mercury._state.weppyConfig.samplingRate,
+				aggregationInterval: Mercury._state.weppyConfig.aggregationInterval
+			});
+			super();
+		}
+
+		track (params: PerfTrackerParams): void {
+			var trackFn = this.tracker;
+
+			if (params.module && typeof params.module === 'string') {
```
Comment (mklucsarits) [inline front/scripts/mercury/modules/Trackers/Perf.ts]: First condition here is not needed if checking typeof.

File: front/scripts/mercury/modules/Trackers/Perf.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+				trackFn = this.tracker.into(params.module);
+			}
+
+			if (params.context) {
```
Comment (mklucsarits) [inline front/scripts/mercury/modules/Trackers/Perf.ts]: Why would the context change during a call to ```track```? Will context always be sent with track? If so, there doesn't seem to be a need to set it initially in the constructor.

File: front/scripts/main/app.ts [hunk 3]
```diff
@@ -14,6 +16,30 @@ var App: any = Em.Application.create({
 	});
 
 App.initializer({
+	name: 'performanceMonitoring',
+	after: 'preload',
```
Comment (rogatty) [inline front/scripts/main/app.ts]: If it's called after `preload` maybe the code should be below it too?

File: front/scripts/mercury/modules/Trackers/GoogleAnalytics.ts [hunk 4]
```diff
@@ -3,19 +3,6 @@ interface Window {
...
-	prefix?: string;
-	// ie. 'UA-32129070-1'
-	id: string;
-	// sampling percentage, from 1 to 100
-	sampleRate: number;
-}
-
-interface GAAccountMap {
-	[name: string]: GAAccount;
-}
```
Comment (rogatty) [inline front/scripts/mercury/modules/Trackers/GoogleAnalytics.ts]: Why did you remove this? What about "atomic PRs"?

File: front/scripts/mercury/modules/Trackers/Perf.ts [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+		constructor () {
+			this.tracker = Weppy.namespace('mercury');
+			this.defaultContext = {
+				skin: 'mercury',
+				'user-agent': window.navigator.userAgent,
+			}
+			this.tracker.setOptions({
+				host: Mercury._state.weppyConfig.host,
+				transport: 'url',
+				context: this.defaultContext,
```
Comment (rogatty) [inline front/scripts/mercury/modules/Trackers/Perf.ts]: Should we add current URL to the context?

File: front/scripts/mercury/modules/Trackers/GoogleAnalytics.ts (hunk 4 above)
Comment (kenkouot) [inline front/scripts/mercury/modules/Trackers/GoogleAnalytics.ts]: I removed this because after I added the WeppyConfig interface to a common location, and sourced it from other files (localSettings.d.ts), it created redundancies with these interfaces. So I removed these interfaces since they are already defined in localSettings.d.ts.

File: front/scripts/mercury/modules/Trackers/Perf.ts (hunk 5 above)
Comment (kenkouot) [inline front/scripts/mercury/modules/Trackers/Perf.ts]: Good idea! Will add that.

File: front/scripts/main/app.ts (hunk 3 above)
Comment (kenkouot) [inline front/scripts/main/app.ts]: Yeah I like that

File: front/scripts/mercury/modules/Trackers/GoogleAnalytics.ts (hunk 4 above)
Comment (kenkouot) [inline front/scripts/mercury/modules/Trackers/GoogleAnalytics.ts]: Added it on different line because it needs to parse URL at time of track, not at instantiation of tracker:
https://github.com/Wikia/mercury/pull/583/files#diff-758befc75e5df0779aa612c8722b5e72R48

File: front/scripts/mercury/modules/Trackers/Perf.ts [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+		track (params: PerfTrackerParams): void {
+			var trackFn = this.tracker;
+
+			if (typeof params.module === 'string') {
+				trackFn = this.tracker.into(params.module);
+			}
+
+			// always set the current URL as part of the context
+			this.defaultContext.url = window.location.href.split('#')[0];
```
Comment (rogatty) [inline front/scripts/mercury/modules/Trackers/Perf.ts]: Will this work if we track something, URL changes and then we track something else? It probably needs flushing after (before?) each transition.

File: front/scripts/mercury/utils/track.ts [hunk 7]
```diff
@@ -160,11 +160,13 @@ module Mercury.Utils {
...
-			var trackerInstance = new trackers[tracker]();
+			var Tracker = trackers[tracker],
+				instance: TrackerInstance;
 
-			if (trackerInstance && trackerInstance.trackPageView) {
+			if (typeof Tracker.prototype.trackPageView === 'function') {
+				instance = new Tracker();
 				Em.Logger.info('Track pageView:', tracker);
-				trackerInstance.trackPageView(trackerInstance.usesAdsContext ? adsContext : context);
+				instance.trackPageView(instance ? adsContext : context);
```
Comment (rogatty) [inline front/scripts/mercury/utils/track.ts]: Seems like a typo. Why check `instance` while calling it?

File: front/scripts/mercury/modules/Trackers/Perf.ts [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+module Mercury.Modules.Trackers {
+	export class Perf extends BaseTracker {
+		public static depsLoaded: boolean = typeof Weppy === 'function';
+		tracker: any;
+		defaultContext: {
+			skin: string;
+			url?: string;
+			'user-agent': string;
```
Comment (wladekb) [inline front/scripts/mercury/modules/Trackers/Perf.ts]: Please add "env" here as we discussed: "prod" for production, something else for other envs incl. preview, verify, devs etc.

File: front/scripts/mercury/modules/Trackers/Perf.ts [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+
+			if (typeof params.module === 'string') {
+				trackFn = this.tracker.into(params.module);
+			}
+
+			// always set the current URL as part of the context
+			this.defaultContext.url = window.location.href.split('#')[0];
+
+			// update context in Weppy with new URL and any explicitly passed overrides for context
+			trackFn.setOptions({
```
Comment (wladekb) [inline front/scripts/mercury/modules/Trackers/Perf.ts]: use trackFn.flush() to trigger immediate send of the queue, but keep in mind all metrics you register afterwards will be bound to the new context

File: server/views/layout.hbs [hunk 10]
```diff
@@ -83,6 +84,11 @@
 			<script src="/front/vendor/i18next/i18next.js"></script>
 			<script src="/front/vendor/vignette/dist/vignette.js"></script>
 			<script src="/front/vendor/scroll-depth/jquery.scrolldepth.js"></script>
+
+			{{#if weppyConfig.enabled}}
```
Comment (hakubo) [inline server/views/layout.hbs]: does it work with
`gulp build --type prod` ?

This run during build:


but this is evaluated during runtime:
{{#if weppyConfig.enabled}}

**Comment by hakubo** [comment]
@kenkouot what are the next steps with this?

File: front/scripts/mercury/modules/Trackers/Perf.ts (hunk 9 above)
Comment (kenkouot) [inline front/scripts/mercury/modules/Trackers/Perf.ts]: @wladekb So should I:
* setOptions -> flush -> (count || store || timer).send ?
* setOptions -> (count || store || timer).send -> flush

Comment (kenkouot) [inline front/scripts/mercury/modules/Trackers/Perf.ts]: Feels like the second option

File: front/scripts/mercury/utils/trackPerf.ts [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+/// <reference path="../modules/Trackers/Perf.ts" />
+'use strict';
+
+/**
+* @description Instantiates performance tracker
+*/
+module Mercury.Utils {
+	var instance: Mercury.Modules.Trackers.Perf;
+	export function trackPerf (obj: PerfTrackerParams) {
+		if (Mercury.Modules.Trackers.Perf.depsLoaded) {
```
Comment (mklucsarits) [inline front/scripts/mercury/utils/trackPerf.ts]: I think it's bad practice to reference a "class" property from another "class" directly, without using a getter. "depsLoaded" should instead be a method (in other words, a getter).

**Comment by mklucsarits** [comment]
Looks mostly OK to me. I just have the one comment. It's up to you if you want to change that or not.
